### ISSUE TITLE
Expand support for common more complex senarios.

### ISSUE THREAD

**Issue by lojack5** [lojack5/structured#2]
Right now this can only handle structs whose size is known at time of code writing.  Implement methods for more dynamic things, like:

 - [x] arrays: Multiple types, whose size are determined by:
   - [x] An unpackable int just prior to the array
   - [x] ~~An attribute on the object (for example, when the size marker doesn't occur just prior to the array.  This one will be a little complicated, as modifying the array should update the size attribute.  Probably need to implement this via a property for the array, or a pre-pack pass that updates the attribute.  Need to think about what to do if two arrays reference the same size attribute, but have different lengths.~~  Skipping this one for now, may  revisit later.  Currently got around this issue by allowing for a size_check option on arrays.
 - [x] binary blobs: currently you can only do `char[13]`, etc for fixed size binary blobs.  Allow for variable sized blobs.  Should support similar options to arrays.
 - [x] ~~Lookahead for dynamic types: this will be more difficult to implement.  Allow for reading future attributes for gathering size information.  Low priority though, since these shouldn't really show up much.~~ Can reopen if needed.
 - [x] Decoding options for strings.  Can probably just implement this with an `unpack_action` on string types, but....
 - [x] More complicated string types:
   - [x] Variable length strings (will basically be handled by binary blobs)
   - [x] More strange string types, like .NET strings.

**Comment by lojack5** [comment]
Groundwork laid in 072a8f2940fb433e84211f4c8d3c9ac63a354dd6, and implemented `array` and `blob` using this.

**Comment by lojack5** [comment]
String types finished in 7b8cb6c54524a625ce8f8186544ae79721ece994

If any of the two skipped options end up being needed, we can reopen:
- [ ] Attribute access for array/char/unicode.
- [ ] Lookahead for array/char/unicode.

**Comment by davidcurie** [comment]
Beautiful work with this project.

I'm not sure if this discussion belongs here or under a request to clarify the [docs](#14).

- [ ] Unpack variable-sized blobs whose size is described by the value from two Lookback values?

Example use case:

```python
class MyStruct(Structured):
    dtype: uint32
    start: uint64  # pointer to start of data
    end: uint64  # pointer to end of data
    data: # bytes from start to end
```

My naive solution would be to try define a Callable that accepts two parameters:

```python
def compute_length(start, end):
    return end - start
```

Then grab `start` and `stop` with `attrgetter` and pass them as arguments in the function to the `LookbackDecider`. The problem I'm seeing with this approach is that I can't make use of a mapping that accepts the result of that length and an additional dtype at the same time. Even if I could make a lookup dictionary that dynamically maps `{(dtype, length) -> MyCustomStruct}`, at that point it seems like making a new class is the proper solution.

I think I'm trying to shoehorn a canned solution of `Header[len, dtype]` where `len` and `dtype` are basic types that describe the length of the array and the size of each array element, respectively. Except in this case, `len` is actually a tuple of `(start, stop)`.

The intuitive case I'd reach for in `struct` would be the following:

```python
struct.unpack(dtype, bstring[start:stop])
```

This looks like something you tried your hand at with subitem 2 in your initial list of array extensions above, or with your Lookahead implementations.

These items are crossed out in your list. In lieu of asking you to reopen them or implement these features, perhaps you could help me and future readers understand how to hook into the low-level mechanics you already implemented.

---

## Subclassing documentation

I'm rusty on the proper way to subclass existing structures for these needs, and how your module keeps track of the remaining bytes to unpack. Is this something you expect to be handled with a custom type

```python
class MyData:
    # Custom implementation
    # Define start, stop, and dtype
    # Not sure which attributes must be defined

class MyStruct(Structured):
    dtype: uint32
    data: Annotated[MyData, SerializeAs(variable)]  # MyData would be one of many possible unions in practice
    # Not sure how to SerializeAs when nested items are not the same type
```

or instead by subclassing a serializer and implementing something like the following?

```python
class MyGenericSerializer(Serializer[uint64, uint64, Generic[T]], init=False):

    def __init__(self, start: uint64, stop: uint64, dtype: Generic[T]) -> None:
        self.start = start
        self.stop = stop
        self.dtype = dtype

    def size(self) -> uint64:
        return self.stop - self.start
      ...
```

Here, the `T` in the generic class depends on the `dtype` from the initial unpacking of `MyStruct` which would be implemented in the `LookbackDecider` and mapped to a proper concrete class like `MyGenericSerializerUint32`.

If serializing is the proper approach for maintaining explicit packing and unpacking, do all of the packing and unpacking methods listed in the API/Serializers section need to be implemented when subclassing `Serializer`? Is this also true if subclassing an existing serializer like `StructActionSerializer`?

**Comment by lojack5** [comment]
Hey, thanks for your interested in this project!

So you could make what's currently in the library work under certain conditions - but only for unpacking:
- If you know the data structure inside the `end - start` bytes at writing time (ie: the data type doesn't depend on other information).
- If you can determine the data structure inside the `end - start` bytes before reading it, in some cases it'd be straightforward to unpack with what's available.

Basically as long as you have a strong understanding of what's inside those bytes, unpack would work fine.  I'd need more information about the data type to write it myself, but the tools are there.

The real issue, and the reason I didn't implement those crossed out items (yet), is writing out your structured data.  There's currently no way to do this cleanly.

Here's an example from [TES4: Oblivion](https://en.uesp.net/wiki/Oblivion_Mod:Mod_File_Format#Records)

```python
record_map = {
    b'ACHR': PlacedNPCRecord,
    b'ACRE': PlacedCreatureRecord,
    # ... and the rest, where PlacedNPCRecord, PlacedCreatureRecord, etc are defined earlier
}

class OblivionRecord(Structured):
    # Record header
    record_type: char[4]
    data_size: uint32
    flags: uint32
    form_id: uint32
    version_control: uint32
    # Record data
    data: Annotated[PlacedNPCRecord | PlacedCreatureRecord | ..., LookbackDecider(attrgetter('record_type'), record_map)]
```

#### Unpacking
If we don't know/haven't implemented the structure for the various records, we might just want to treat that record as raw bytes for now.  Currently we don't have a way to do this, without (as you noted) implementing our map as a custom classed dictionary that returns the `char[data_size]` serializer on accessing that unkown/unimplemented record. This is pretty much a showstopper at this point.

For the other record types, we're relying on the fact that we know the detailed structured of the records, so we don't really need the size to unpack them.  We could use `__post_init__` to *verify* that the data we unpacked matches in size to the `data_size` field.

#### Packing
Here we're pretty stuck.  Currently there are no built in tools in `structured` to tell the serializer to "rewind" and write the `data_size` field.  Since the records could have various final sized (even amongst the same record types), this needs to be done after writing the record data.

Currently, this idea of rewinding to write a size is *only* implemented as a special case for the `array[Header[len, size], type]` serializer.

#### TL;DR
Unpacking you can probably handle, but writing is an issue.  This stems from coming up with a generic way of telling the serialize to rewind to a specific point in the stream/bytes/file to write a size.  Moreover, in your case we need this implementation to be generic enough to not just write a size, but two values whose difference is the size.

This is a relatively hard problem, because there's a lot to think about as far as API design goes once we open up to that:
1. Writing something other than just the size means we want to allow the user to provide the data to be written.  The structure of this information isn't hard to allow the user the provide (in your case, a `tuple[uint64, uint64]`).
2. The user needs to provide the values to be written - there are a lot of possibilities here in the generic case.  The values might be an offset relative to the start of the file, the start of this particular `Structured` subclass's data, an absolute position in the file/bytestream, etc.  So what information do we send to the user to determine this?  The more information we want to send, the more additional tracking we need to add to the code (keeping track of offsets to each attribute, plus some).


Finally to your question about Serializers - the serializer methods marked as "not required" don't need to be implemented, as they have default implementations that should work in most cases.  Otherwise yes, the rest of the methods and attributes need to be implemented.  Note, `size` is an attribute, so you need to either implement it as a `property`, or dynamically update it (for an example, see the [Union serializer](https://github.com/lojack5/structured/blob/main/structured/serializers/unions.py#L62).  I'll also note that this `size` attribute isn't a value "telling" `Structured` how many bytes to read/write - the `Structured` class itself isn't the one responsible for this, it's the Serializer itself that does the reading of the bytes before unpacking.  The `size` attribute is merely an "after the fact" accounting of how many bytes the serializer read/write in doing its job.

Long winded response to say - you can probably get unpacking to work fine (I'd need detailed information about the data structure to give you more hints), but packing is a problem.  To implement something for writing, some feedback on how you'd think that should look would be welcome!

**Comment by davidcurie** [comment]
Thanks for such a detailed explanation.

I was beginning to suspect problems with packing up the data when I read more about how serializers work and how to write my own `pack` functions. Glad to have the specific details of the problem areas identified. Making a generic catch-all serializer is difficult.

I can't speak to how one should extend this in general, but I can speak to the specific needs in my case. I'm happy to solve the problem for my unique needs and share a solution for you to post in the wiki demonstrating how one could extend the API. Having a tangible example to study with context would have helped me understand the API a little more.

In my case, the data structure is fixed and known ahead of time during unpacking because of the preceding fields. The `(start, stop)` tuple in the header refers to the absolute byte position from the top of the file. The `start` byte references the position immediately after the subsequent `stop` pointer. There is effectively only one array per data pointer, except the unpacked array could itself be another self-contained `header:array` byte string. This is iterated by recursion. If you're curious, a detailed description of the unpacking algorithm can be found in a [similar Matlab project for opening instrument files](https://github.com/ElsevierSoftwareX/SOFTX-D-20-00088/blob/master/%2BWITio/%2Bdoc/README%20on%20WIT-tag%20format.txt) from a particular microscope.

I gathered that I could boil my template down to something like the following:

```python
dtype_map = {
    0: InstrumentField  # TypeVar proxy or Self, meant to recurse the object unpacking
    ⋮
    9: unicode[uint32]
}

class BundledTag(Structured):
    dtype: uint32
    start: uint64
    stop: uint64
    data: Annotated[Self | uint16 | ..., LookbackDecider(attrgetter("dtype"), dtype_map)]

class InstrumentField(Structured):
    key: unicode[uint32]
    value: BundledTag
```

The result of an `InstrumentField` object should be a nested dictionary of arbitrary depth, although anecdotally this is not greater than 5 or 6. The deepest nested dictionaries hold images (RGB arrays), text comments, or other flags. What I care about is the nested `{key: value}` pairs, but what I get with this unpacking is some extra fields because the `BundledTag` is spitting out extra values. It rightly thinks I want to unpack `dtype`, `start`, and `stop` values from my byte string because I define them in a Structured class. I can write `@property` attributes to make the data fields accessible as the attributes I want. Importantly, because these `start`, `stop` fields exist during my unpacking, they exist during packing, so the file succeeds in a read/write pipeline to first order while maintaining original file compatibility.

In practice, I run into two problems.

1. I don't know how to recursively iterate when `dtype = 0`. Passing in `typing.Self` fails because it doesn't recognize itself as a serializable type. Passing in a TypeVar proxy for the class also gives me the same error.
    - I can solve this locally by defining a series of `Structured` classes (one for each depth of the nested dictionary) and calling them sequentially. This feels redundant, but I can confirm that the fields are unpacked with this structure.

2. If I append any data to existing fields, I not only need to update the `end` pointer for the local field, but the subsequent `start` and `end` pointers for the remaining fields that follow.
    - This problem goes away if I pinky promise not to touch the data length of the internal arrays, meaning it is safe to amend but not safe to append.

Problem 2 helps me understand the main crux of the issue; keeping track of the current byte position during the pack and unpack operations is not implemented.

If I relax the need to be able to pack the data afterwards and I'm understanding your library properly, I think I can make a specific serializer for my unpacking needs. Instead of manually mapping the `data` results of a `BundledTag` object to each instance of a specific `InstrumentField`, I could just define that the `value` field from `InstrumentField` should be unpacked as an object according to `BundledTagSerializer`. I think this will look something like the existing `BundledTag`, and after your clarifying remarks and increased study of your codebase I now know that I'd need to implement my own `pack` method and redefine the `size` property.

My confusion stems from how to translate the serializer information in your documentation to the real world. Is it better for me to extend or define `StructSerializer[int, int, int]` to accept `start, stop, dtype` variables? Should I make a modified version of `StaticStructArraySerializer`? If the latter, then I'm lost on how I should control the arguments to my new class. If I know where the `__init__` function is called, I think my problem is solved by either replacing `count` with separate `start, stop` arguments, and then defining `self.count = stop - start`, or by allowing count to be a tuple (maintains the same number of arguments) and redefining `self.count = count[1] - count[0]`. Once I get my serializer defined, is it as simple as dropping it in for the type hint to the value?

```python
class InstrumentField(Structured):
    key: unicode[uint32]
    value: BundledTagSerializer  # ?
```

**Comment by lojack5** [comment]
Thanks for the more in depth information!  So from what I'm seeing, at the very least a few features could be added as stop-gaps while I think on the "write a serialized attribute that could be in some arbitrary position in the serialization order but depends on information later in the process" problem.

- [x] `typing.Self` as a serializable type-hint.  When I started this, Python 3.11 wasn't even out yet (nor even in RC stage).  And of course allow back-compatability via `typing-extensions`.

I tried to think of some sort of tricky way you could accomplish a `Self`-like type-hinting with what's currently available, but I'm fairly certain nothing would work currently.  And writing your own Serializer class to handle this is doable (with s a bit of hackery), it's be better suited in the library proper (and useful for other things too).  It certainly opens up the possibility of recursion which can enable all sorts of cool things.

Indeed as you mentioned - Problem 2 is the "way" to handle it currently - if you're careful not to affect the sizes, it'll work out.  Not a satisfying answer though.

Here's a rough draft of an idea I've been thinking on for a generic way to handle "rewinding" to write attributes later:

```python
class ComplexObject(Structured):
    some_data: uint32
    needs_later_info: Annotated[uint16, Delayed(some_callable, 'data_that_was_needed')]
    some_other_data: unicode[100]
    data_that_was_needed: OtherComplexStructuredData
```
The idea is including a `Delayed` object in the `Annotated` annotation would signal that this attribute doesn't have the information to write it properly until some later attribute is written.  Details might change, but the 2 parameters there are:
- A callable to be called when ready to write.  The callable would take the Structured object to be serialized (and possible more information) and return the value that should be used for `needs_later_info`.
- An attribute name (`'data that was needed'`) that tells the Serializer after which attribute is written it's safe to go back and write this attribute.

Still some details to work out though (this is more notes for me), since I could see this being used for many senarios:
- I need a better way to track "position" in a byte-stream/file object.  Due to the recursive and composing nature of Serializers, this could be tricky.
- What additional info needs to be passed to the callable? Sizes of all previously serialized attributes?  Absolute positions of attributes? Offsets of attributes?

Just an example of how a simple problem can end up with some surprisingly nested serializers, [Oblivion shader packages](https://github.com/wrye-bash/oblivionsdp/blob/structured/shaderpackage.py).  In this example we have:

```
                   ShaderPackage (CompoundSerializer)
                         |
StructSerializer(uint32) + ArraySerializer(for Shader)
                                |
               HeaderSerializer + repeated use of StructuredSerializer[Shader]
                                                         |
                                                 CompoundSerializer (the one built for Shader)
                                                         |
                                       UnicodeSerializer + StructSerializer(uint32)
                                              |                   
                                     DynamicCharSerializer
```
So tracking even relative offset can get tricky.  Also opens up the question, "relative to where?".  The `CompoundSerializer` for `Shader` doesn't know it's going to be put inside another serializer, so it's serializer *should* just track offsets relative to the beginning of that particular `Shader` object, for example.  Now as soon as you throw absolute positions in the mix, it adds even more complexity to the underlying machinery to get it right no matter the situation.

Not saying I won't do it, just that it may take me quite a while to get a solution that works correctly in all possible usages.

BTW thanks for the link to the file format you're dealing with specifically, let me really dig in to see if there was any possibility of using what's already there in a unique way to get results.

Anyway, I'll get working on `typing.Self` handling at the very least soon-ish (I want to put finishing touches on the "Consume until end" variant of array serializers that was requested first).
